tests/catch/unit/memory/hipFree.cc

```diff
@@ -0,0 +1,483 @@
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+
+#include "DriverContext.hh"
+#include "MemUtils.hh"
+
+/*
+ * This testcase verifies [ hipFree || hipFreeArray || hipArrayDestroy || hipHostFree with
+ * hipHostMalloc ]
+ * 1. Check that hipFree implicitly synchronises the device.
+ * 2.Perform multiple allocations and then call hipFree on each pointer concurrently (from unique
+ * threads) for different memory types and different allocation sizes.
+ * 3. Pass nullptr as argument and check that no operation is performed and hipSuccess is returned.
+ * 4. Pass an invalid ptr and check that hipErrorInvalidValue is returned.
+ * 5. Call hipFree twice on the same pointer and check that the implementation handles the second
+ * call correctly.
+ * 6. HipHostFree only:
+ *    Try to free memory that has been registered with hipHostRegister and check that
+ * hipErrorInvalidValue is returned.
+ */
+
+
+enum FreeType { DevFree, ArrayFree, ArrayDestroy, HostFree };
+
+// helper functions to release memory
+template <typename T> hipError_t freeStuff(T ptr, FreeType type) {
+  switch (type) {
+    case DevFree:
+      return hipFree(ptr);
+      break;
+    case HostFree:
+      return hipHostFree(ptr);
+      break;
+    case ArrayFree: {
+      auto arrPtr = reinterpret_cast<hipArray_t>(ptr);
+      return hipFreeArray(arrPtr);
+      break;
+    }
+    case ArrayDestroy: {
+#if HT_NVIDIA
+      auto arrPtr = reinterpret_cast<hiparray>(ptr);
+      return hipArrayDestroy(arrPtr);
+#else
+      auto arrPtr = reinterpret_cast<hipArray_t>(ptr);
+      return hipFreeArray(arrPtr);
+#endif
+      break;
+    }
+    default:
+      return hipErrorIllegalState;
+      break;
+  }
+}
+
+// Helper function to check if work on device is done
+template <typename T> hipError_t workIsDoneCheck(T ptr, FreeType fType) {
+  auto error = hipSuccess;
+  // free memory
+  error = freeStuff(ptr, fType);
+  if (error != hipSuccess) {
+    printf("Breaking on free \n");
```

Comment:
There are some macros for occasion like this, such as WARN or FAIL https://github.com/catchorg/Catch2/blob/devel/docs/logging.md